File: Android/app/src/main/java/com/peaa/LazyLoad/LazyLoadDomainObject.java
package com.peaa.LazyLoad;

import com.peaa.DataManipulation.BaseMapperInterfaces.IBaseMapper;
import com.peaa.Domain.DomainObject;
import com.peaa.Interfaces.ILazyLoadDomainObject;
import com.peaa.Interfaces.ILazyLoader;

/**
 * Created by aiko on 4/2/17.
 */

public abstract class LazyLoadDomainObject<TSearchInput> extends DomainObject implements ILazyLoadDomainObject<TSearchInput> {
    TSearchInput _criteria;

    public TSearchInput GetCriteria (){return _criteria;}

    public LazyLoadDomainObject(TSearchInput criteria, IBaseMapper mapper) {
        super(mapper);

        _criteria = criteria;
    }

    public <TEntity extends  LazyLoadDomainObject<TSearchInput>> void Load(ILazyLoader<TEntity, TSearchInput> loader)
    {
        if ((_criteria == null) || (loader == null))
            return;

        TEntity currentEntity = (TEntity) this;

        loader.LoadAllFields(currentEntity, _criteria);

        _criteria = null;
    }
}

File: DotNet/CSharp/Framework/LazyLoad/LazyLoadDomainObject.cs
using Framework.Data_Manipulation;
using Framework.Domain;

namespace Framework.LazyLoad
{
    public interface ILazyLoadDomainObject<TSearchInput>
    {
        TSearchInput Criteria { get; }
        void Load<TEntity>(ILazyLoader<TEntity, TSearchInput> loader)
           where TEntity : LazyLoadDomainObject<TSearchInput>;
    }

    public abstract class LazyLoadDomainObject<TSearchInput> : DomainObject, ILazyLoadDomainObject<TSearchInput>
    {
        private TSearchInput _criteria;

        public TSearchInput Criteria { get { return _criteria; } }

        public LazyLoadDomainObject(
            TSearchInput criteria,
            IBaseMapper mapper
            ) : base(mapper)
        {
            _criteria = criteria;
        }

        public void Load<TEntity>(ILazyLoader<TEntity, TSearchInput> loader)
            where TEntity : LazyLoadDomainObject<TSearchInput>
        {
            if ((_criteria == null) || (loader == null))
                return;

            TEntity currentEntity = (TEntity) this;

            loader.LoadAllFields(ref currentEntity, _criteria);

            _criteria = default(TSearchInput);      //Clear criteria
        }
    }
}
